cmake/CompileSlice.cmake

```diff
@@ -0,0 +1,20 @@
+# CompileSlice - script slice compile with cmake similar to IceBuilder
+# 
+# Params:
+# slice2bin_params - parameters passed for specific slice2_bin use cases
+# slice_include_paths - include dir(s) needed for -I parameter
+# ice_file - file to be compiled (<filename>.ice)
+# output_dir - generated source output path
+# output_file - generated source output file name
+
+function(CompileSlice slice2bin_params slice_include_paths ice_file output_dir output_file)
```

Comment:
The function only creates slices from ice files it is given and the expected output file. If the compiles sources are needed by other projects then it just needs this:

`target_include_directories(<target> PRIVATE ${GENERATED_SLICES_DIR})`

Reply:
I mean with other build systems we use the output of `slice2cpp --depend` to compute what other Slice files need to be rebuild when a given file changes. If you update `BuiltinSequences.ice` then `PropertiesAdmin.ice` must also be rebuild because the later includes the former.

Reply:
I have not tested that particular feature, but I believe since the ice file is the `MAIN_DEPENDENCY` that if the contents of that file change it will trigger a rebuild.

https://cmake.org/cmake/help/v3.7/command/add_custom_command.html

Reply:
my understanding here is that `MAIN_DEPENDENCY` is set to `ice_file` so if the file change it gets rebuild, but if any of the files included by it changes it doesn't. The rule doesn't add any other dependencies so it doesn't know what other files a Slice file depends on. 

I assume this runs once per Slice file in the set, and only the file modified gets rebuild.

Reply:
I could check that this evening. If `--depend` needs to be set I can do it here:

https://github.com/ZeroAbility/ice/blob/ad731868b014e628b062d4eaa71c0866e81d5f7c/cpp/src/Ice/CMakeLists.txt#L405

Reply:
I was able to add another command to this function that creates a `txt` file for each `ice` file. That command will also depend on the `ice` file so that it will recreate the `txt` file if there is a change. My thought was to convert the `txt` file to a variable using [`file(STRINGS)`](https://cmake.org/cmake/help/v3.7/command/file.html?highlight=STRINGS) and iterating over the paths in the file to pass to the slice compiler. Is there a particular order the build needs to happen (i.e. file -> dependees, or dependees -> file)?

Thank you

Reply:
First write the dependencies, then the generated code. I think we use `.ice.d` extension for the dependency files with other build systems.

The format of --depend is Makefile style depdencies like:
```
Locator.h: \
 ..\..\slice\Ice\Locator.ice \
 ..\..\slice/Ice/Identity.ice \
 ..\..\slice/Ice/Process.ice
```
Will CMake understand this or do you need to post-process it?

Reply:
I am able to remove extra chars from the file and assign it to a var in CMake. I may be able to use [AddFileDependencies](https://cmake.org/cmake/help/v3.7/module/AddFileDependencies.html) to achieve this. Could you provide me with some test cases and expectations so I can make sure it is working properly? Thank you.

Reply:
In the above example from Locator.ice, you should see that modifications to Identity.ice trigger rebuild of Locator.ice or other files that directly or indirectly include Identity.ice

Reply:
I'm still looking into this. CMake looks like it should be able to handle this but I have not had success with it yet. Could you refer me to the script or code that normally handles this so I can compare it to how I am trying to implement it with CMake?
Thank you,

Reply:
You basically just need to check that all dependencies are older than the generated file, or rebuild it. Make figure this automatically from the above --depend output.

For MSBuild we have a custom task that parses the output of --depend-xml, is similar to --depend but in XML format for easy parsing.

https://github.com/zeroc-ice/ice-builder-msbuild/blob/master/Tasks.cs#L713

most interesting part in

https://github.com/zeroc-ice/ice-builder-msbuild/blob/master/Tasks.cs#L868

For other build systmem that don't handle Make file depdencies automatically, we have similar scripts

https://github.com/zeroc-ice/ice-builder-gradle/blob/master/src/main/groovy/com/zeroc/gradle/icebuilder/slice/SliceTask.groovy#L485

Reply:
I'm still looking into performing this task in CMake so there doesn't have to be several types of processes to achieve the same thing. The current function revision is not working.

Reply:
I've come to a sensible compromise given the limitations of CMake and the requirements of the ice files. The most recent push has a simplified function that is called during the Configure stage and will trigger a rebuild of the configuration if an ice file is changed. This will cause all of the ice files to regenerate the source files. Trying to do this on a case by case basis seemed not to work consistently if at all. I have tested the following generators and it appears to produce the desired behavior:

- Visual Studio 16 2019 and ninja, Windows x64
- Unix Makefiles and ninja, Debian Buster WSL, Linux x64

Please let me know if this would suffice.

Reply:
We'll give it a try, I think it should be ok. And we can always improve this Slice dependencies handling later on.

Could you update the main cpp README or add a new README with a description of this new cmake build system - things like the various steps to build Ice C++ with cmake, and options like "build just the source", "build the source and tests", "install" (if provided), build debug / build release...

Also, did you look into factoring the boilerplate code included in lots of the new cmake files (especially the tests)? It would be nice to move this boilerplate to a common include file.

Reply:
If you end up testing the dependencies let me know if there is any issues. I didn't see anything that stood out to me. 

I was going to update/add README documentation after sorting out the tests. It should be fairly straightforward. The same dependencies are still needed, and I will include a blurb about taking advantage of vcpkg to deliver them. I will also include links to the CMake documentation for the defined `find_package` statements. If there is someone there that could test this with the nuget packages let me know. Otherwise, I can install them and test it. Installation is handled by [CMAKE_INSTALL_PREFIX](https://cmake.org/cmake/help/v3.7/variable/CMAKE_INSTALL_PREFIX.html) and can be modified by the developer as needed. Otherwise, it has default paths specified. All the targets have an `install()` specified. 

I looked into creating functions that would add targets given the specific details for each test with similar contents in the CMakeLists. That is probably the most efficient way to achieve the goal of reducing the amount of duplication. Each target has to be unique but the output files can be anything (i.e. Client, Server, etc...). If there is a specific pattern you would like me to use for the target names, please let me know.